lib/puppetserver/ca/action/prune.rb

```diff
@@ -0,0 +1,123 @@
+require 'optparse'
+require 'openssl'
+require 'puppetserver/ca/errors'
+require 'puppetserver/ca/utils/cli_parsing'
+require 'puppetserver/ca/utils/file_system'
+require 'puppetserver/ca/utils/config'
+require 'puppetserver/ca/x509_loader'
+
+module Puppetserver
+  module Ca
+    module Action
+      class Prune
+        include Puppetserver::Ca::Utils
+
+        SUMMARY = "Prune the local CRL on disk to remove any duplicated certificates"
+        BANNER = <<-BANNER
+Usage:
+  puppetserver ca prune [--help]
+  puppetserver ca prune [--config]
+
+Description:
+  Prune the list of revoked certificates of any duplication within it.
+
+Options:
+BANNER
+
+        def initialize(logger)
+          @logger = logger
+        end
+
+        def self.parser(parsed = {})
+          OptionParser.new do |opts|
+            opts.banner = BANNER # I will get a banner soon
+            opts.on('--help', 'Display this command-specific help output') do |help|
+              parsed['help'] = true
+            end
+            opts.on('--config CONF', 'Path to the puppet.conf file on disk') do |conf|
+              parsed['config'] = conf
+            end
+          end
+        end
+
+        # TODO:
+        # 1. Ensure that everything is offline (Done)
+        # 2. Get the path to the CRL (Done)
+        # 3. Prune each issuer revoked list (In progress)
+        # 4. Addin debug output
+        # 5. Repack the CRL and write it out?
+
+        def run(inputs)
+          config_path = inputs['config']
+
+
+          # Validate the config path.
+          if config_path
+            errors = FileSystem.validate_file_paths(config_path)
+            return 1 if Errors.handle_with_usage(@logger, errors)
+          end
+
+          # Validate puppet config setting.
+          settings_overrides = {}
+          puppet = Config::Puppet.new(config_path)
+          puppet.load(settings_overrides)
+          return 1 if Errors.handle_with_usage(@logger, puppet.errors)
+
+          # Validate that we are offline
+          return 1 if HttpClient.check_server_online(puppet.settings, @logger)
+
+          # Getting the CRL(s)
+          loader = X509Loader.new(puppet.settings[:cacert], puppet.settings[:cakey], puppet.settings[:cacrl])
+
+          crl_list = loader.crls
+          prune_per_issuer(crl_list)  # Prune the list
+          return 0                  # Place holder return value for now
+        end
+
+        # Further note: The class Revoked override the == method.  We could make
+        # use of this by using array.uniq method to dedup ?
```

Comment:
Not sure what you're referring to here..

Reply:
Ooh, the documentation for the class X509::Revoked has the method `==` override for comparison between 2 Revoked object. Spent sometime researching and found out that the Array class method `uniq` utilize the `==` method so I thought it could be another way of pruning the array.  Just some note I left in case performance aren't good 😄

Reply:
that'd be a pretty easy implementation ( :

Reply:
I'll spend sometime digging around on that.  I'm still learning the behind the scene of Ruby :)

Reply:
So I got this to work for pruning:
```
revoked_list.uniq! do |rev|
  rev.serial
end
```
But haven't figure out how to slide the debugging function for the logger into that block for only the removed cert.  Any good idea on how to do it?

Reply:
I forgot about that requirement. I doubt you can log for the removed revocation with that method. Unless you iterated over the list twice to compare what did exist to what no-longer does.

Reply:
Yeah I got stuck on trying to slip the debug function in, couldn't figure out how to do it without having to output info for the whole array.  I'll stick with the current implementation for now and maybe compare their performance.